File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-6.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        hs = set()
        maxL = 0
        l, r = 0, 0

        while r < len(s):
            print(hs)
            while s[r] in hs:
                hs.remove(s[l])
                l += 1

            hs.add(s[r])
            maxL = max(maxL, r - l + 1)
            r += 1

        return maxL

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-6.js
class Solution {
    /**
     * @param {string} s
     * @param {number} k
     * @return {number}
     */
    characterReplacement(s, k) {
        const hm = new Map()
        let len = 0
        let l = 0

        for (let r = 0; r < s.length; r++) {
            hm.set(s[r], 1 + (hm.get(s[r]) || 0))
            
            let curMax = Math.max(...hm.values())

            while ((r - l + 1) - (curMax) > k) {
                hm.set(s[l], hm.get(s[l]) - 1)
                l++

                curMax = Math.max(...hm.values())
            }
            len = Math.max(len, r - l + 1)
        }

        return len
    }
}
